internal/driver/flamegraph.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package driver
+
+import (
+	"encoding/json"
+	"html/template"
+	"net/http"
+
+	"github.com/google/pprof/internal/report"
+)
+
+type flameGraphNode struct {
+	Name     string
+	Value    int64
+	Children map[string]*flameGraphNode
+}
+
+func (n *flameGraphNode) add(stackPtr *[]string, index int, value int64) {
+	n.Value += value
+	if index >= 0 {
+		head := (*stackPtr)[index]
+		childPtr, ok := n.Children[head]
+		if !ok {
+			childPtr = &(flameGraphNode{head, 0, make(map[string]*flameGraphNode)})
+			n.Children[head] = childPtr
+		}
+		childPtr.add(stackPtr, index-1, value)
+	}
+}
+
+func (n *flameGraphNode) MarshalJSON() ([]byte, error) {
+	v := make([]flameGraphNode, 0, len(n.Children))
+	for _, value := range n.Children {
+		v = append(v, *value)
+	}
+
+	return json.Marshal(&struct {
+		Name     string           `json:"n"`
+		Value    int64            `json:"v"`
+		Children []flameGraphNode `json:"c"`
+	}{
+		Name:     n.Name,
+		Value:    n.Value,
+		Children: v,
+	})
+}
+
+// flamegraph generates a web page containing a flamegraph.
+func (ui *webInterface) flamegraph(w http.ResponseWriter, req *http.Request) {
+	rpt, errList := ui.makeReport(w, req, []string{"svg"})
+	if rpt == nil {
+		return // error already reported
+	}
+
+	// Obtaining prof from report
+	prof := rpt.Prof()
+
+	// Get sample_index from variables
+	si := pprofVariables["sample_index"].value
+
+	// Getting default index
+	index, err := ui.prof.SampleIndexByName(si)
```

Comment:
I think that we could simplify the code for the flame graph data generation even more if we'd use the existing graph generation call here - https://github.com/google/pprof/blob/f0300aaaf05b5fe9bec4a79ff111f9578c03e533/internal/report/report.go#L236

It has the mode to generate a tree, but that is guarded by options right now. We could factor out a separate function for forceful generation of the graph in the tree format and generate the flame graph data out of that.

The great thing about that is that we would get the sample value selection for free I think, as well as unit formatting and as well as the granularity options would work - for example, running "pprof -files -http :8080" and going to the flame graph view would display the flame graph in terms of source files granularity. Similarly, using "-lines" instead of "-files" would display the line numbers in the labels.

Reply:
Possibly. Haven't tried the tree format. The graph format doesn't have enough information for the flame graph to be assembled. If you have a data structure example of the tree output, I could have a look.

Reply:
@spiermar Something like https://github.com/google/pprof/compare/master...aalexand:aalexand-flamegraph?expand=1 is what tried and it seems to work.

Reply:
We should also return the total value and use that for formatting the percentage rather than compute it based on the flame graph node width - this is to make sure that when a focus filter is applied the percentage is still relative to the unfocused total unless relative_percentages option is set on the command line.

Reply:
You mean keep the parent node sizes the same, but hide the nodes that filtered out? If that's the case, I would probably need an "original value"  or "total value" in the sample. Is that available? Haven't checked the structure yet. 
As for the tree implementation, I still need to get some time to check it. The main reason why I stayed away from the graph was because I couldn't extract all the information I needed for the flame graph, specially on cases where multiple edges joined in a vertex. If that is not a problem with the tree, and this implementation is clearer for the the maintainers, should be good.

Reply:
@spiermar No, the size of the nodes on the screen would be like today, I am only talking about the percentage value shown in the node text and tooltips.

Regarding the tree, when the call_tree flag is passed to the graph generation like I'm doing in the code I linked, the graph is actually a tree so there is at most one incoming edge for each node (the parent link).

Reply:
@spiermap Does using the additional values in the flame graph nodes to get access to them in the formatting functions look good to you in general? This relies on that the values being passed around as is and I don't know if it works-as-designed or works-as-implemented.

Reply:
That should be simple, as long as I can get the total value, the plugin has an option to pass a "label" function, that returns the value and format of the label, and the tooltip. The `call_tree` flag option looks good. I just wasn't aware of that option in the beginning.